File: LC/2265. Count Nodes Equal to Average of Subtree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.ans = 0

    def f(self, root):
        if not root:
            return (0, 0)
        l = self.f(root.left)
        r = self.f(root.right)
        s = l[1] + r[1] + root.val
        n = 1 + l[0] + r[0]
        self.ans += int(s / n) == root.val
        return (n, s)

    def averageOfSubtree(self, root: TreeNode) -> int:
        self.f(root)
        return self.ans

File: LC/2265. Count Nodes Equal to Average of Subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int ans = 0;
    pair<int,int> f(TreeNode* root) {
        if(!root) return {0,0};
        auto l = f(root->left);
        auto r = f(root->right);
        int s = l.second + r.second + root->val;
        int n = 1 + l.first + r.first;
        ans += (s / n) == root->val;
        return {n,s};
    }
public:
    int averageOfSubtree(TreeNode* root) {
        f(root);
        return ans;
    }
};